xmtp_mls/src/groups/device_sync.rs

```diff
@@ -399,540 +358,514 @@ where
+        acknowledge: F,
+        handle: &WorkerHandle<SyncMetric>,
+        retry: &Retry,
+    ) -> Result<(), DeviceSyncError>
+    where
+        F: Fn() -> Fut,
+        Fut: std::future::Future<Output = Result<(), DeviceSyncError>>,
+    {
+        tracing::info!("Sending sync payload.");
+        let provider = Arc::new(self.mls_provider()?);
+
+        match acknowledge().await {
+            Err(DeviceSyncError::AlreadyAcknowledged) => {
+                return Ok(());
+            }
+            result => result?,
         }
 
-        Ok(())
-    }
+        let Some(device_sync_server_url) = &self.device_sync.server_url else {
+            tracing::info!("Unable to send sync payload - no sync server url present.");
+            return Err(DeviceSyncError::MissingSyncServerUrl);
+        };
 
-    async fn create_sync_reply(
-        &self,
-        request_id: &str,
-        syncables: &[Vec<Syncable>],
-        kind: DeviceSyncKind,
-    ) -> Result<DeviceSyncReplyProto, DeviceSyncError> {
-        let (payload, enc_key) = encrypt_syncables(syncables)?;
-
-        // upload the payload
-        let Some(url) = &self.history_sync_url else {
-            return Err(DeviceSyncError::MissingHistorySyncUrl);
+        let mut request_id = "".to_string();
+        let options = if let Some(request) = request {
+            let Some(options) = request.options else {
+                return Err(DeviceSyncError::MissingOptions);
+            };
+            request_id = request.request_id;
+            options
+        } else {
+            default_backup_options()
         };
-        let upload_url = format!("{url}/upload");
-        tracing::info!(
-            inbox_id = self.inbox_id(),
-            installation_id = hex::encode(self.installation_public_key()),
-            "Using upload url {upload_url}",
-        );
 
-        let response = reqwest::Client::new()
-            .post(upload_url)
-            .body(payload)
-            .send()
-            .await?;
+        // Generate a random encryption key
+        let key = xmtp_common::rand_vec::<32>();
+
+        // Now we want to create an encrypted stream from our database to the history server.
+        //
+        // 1. Build the exporter
+        let exporter = BackupExporter::new(options, &provider, &key);
+        let metadata = exporter.metadata().clone();
+
+        #[cfg(not(target_arch = "wasm32"))]
+        let body = {
+            // 2. A compat layer to have futures AsyncRead play nice with tokio's AsyncRead
+            let exporter_compat = tokio_util::compat::FuturesAsyncReadCompatExt::compat(exporter);
+            // 3. Add a stream layer over the async read
+            let stream = tokio_util::io::ReaderStream::new(exporter_compat);
+            // 4. Pipe that stream as the body to the request to the history server
+            reqwest::Body::wrap_stream(stream)
+        };
+        #[cfg(target_arch = "wasm32")]
+        let body = {
```

Comment:
This is probably fine for now, but it would be nice to separate platform-specific code out into traits, for instance a `DeviceSyncBody` trait that can create a Wasm body or a native body. that way as our experience & understanding of WebAssembly improves, we can modify the body as needed w/o modifying the device sync code. For instance, i could see us improving this code by using web_sys apis to get something resembling a stream in the future. 

its also nice not to worry about #[cfg] in code itself which makes it a bit less readable imo

a native vs wasm request body could also be useful elsewhere, like to other potential workers or something

would be a good followup for the libxmtp board